### PR TITLE
refactor config error handling

### PR DESCRIPTION
**Pull request by vankichi**
### Description:

I refactored the error handling in each `internal/config/` file.
When some error occurs, the `Opts()` method will return `(nil, error)` instead of `([]some.Options, error)`.
In addition, I updated tests and logics which are affected by the above changes.

### Environment:

- Go Version: 1.16
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by hlts2** [comment]
thank you for pr.
Should we fix the following error?
https://github.com/vdaas/vald/pull/1190/checks?check_run_id=2376708222

**Comment by kevindiu** [comment]
can you please fix this build error?
https://github.com/vdaas/vald/pull/1190/checks?check_run_id=2376983056

File: hack/benchmark/internal/client/ngtd/grpc/option.go [hunk 1]
```diff
@@ -27,17 +27,20 @@ type Option func(*ngtdClient)
...
-			},
-			CallOption: &config.CallOption{
-				MaxRecvMsgSize: 100000000000,
-			},
-			DialOption: &config.DialOption{
-				Insecure: true,
-			},
-		}).Bind().Opts()...,
+		func() (opts []grpc.Option) {
+			opts, _ = (&config.GRPCClient{
```
Comment (hlts2) [inline hack/benchmark/internal/client/ngtd/grpc/option.go:31]: Should we output the error log to notify the user?

**Review by hlts2 — APPROVED** [review]

File: hack/benchmark/internal/client/ngtd/grpc/option.go [hunk 2]
```diff
@@ -27,17 +27,20 @@ type Option func(*ngtdClient)
...
+				Addrs: []string{
+					"127.0.0.1:8200",
+				},
+				CallOption: &config.CallOption{
+					MaxRecvMsgSize: 100000000000,
+				},
+				DialOption: &config.DialOption{
+					Insecure: true,
+				},
+			}).Bind().Opts()
```
Comment (kevindiu) [inline hack/benchmark/internal/client/ngtd/grpc/option.go:41]:
```suggestion
			c := &config.GRPCClient{
				Addrs: []string{
					"127.0.0.1:8200",
				},
				CallOption: &config.CallOption{
					MaxRecvMsgSize: 100000000000,
				},
				DialOption: &config.DialOption{
					Insecure: true,
				},
			}
			opts, err := c.Bind().Opts()
			if err != nil {
				log.Error(err)
			}
```

**Review by kevindiu — APPROVED** [review]
LGTM,
If you think it requires author review, please ask them after @hlts2 complete the review

**Review by hlts2 — APPROVED** [review]

**Comment by vankichi** [comment]
@kpango @rinx 
I got the LGTM from 2 SET members and this PR may require authors' review.
Please review this PR by 4/30(Fri)?
If this schedule is tough, I re-define the review deadline.
thank you.

**Comment by rinx** [comment]
the build of the whole project is failing. Please check it.
https://github.com/vdaas/vald/pull/1190/checks?check_run_id=2444147894

the internal test is also failing, but it seems not caused by this PR...?
https://github.com/vdaas/vald/pull/1190/checks?check_run_id=2444148891

**Comment by vankichi** [comment]
@rinx 
thanks for checking.
I think both of these failures are not caused by this PR.
I'll confirm these failures again and notice you later.

**Comment by kevindiu** [comment]
for the first error, we need to import the log package due to this changes in this PR.
https://github.com/vdaas/vald/pull/1190/files#diff-67ecb51fef0ac9bb479bf47134a52cbc2375c3c2d88add06170e948b8251803bR44

for the  second error,  maybe restart the ci will fix the problem

**Comment by vankichi** [comment]
@rinx 
sorry for the late announcement.
I checked the failed CI log and re-run CI jobs.
About unit testing, it ends with success.
On the other hand, e2e bench still ends with failure.
This failure seems to not elated this PR. Which is the best to fix this failure in this PR or create a new PR to fix this failure?

File: internal/config/net.go [hunk 3]
```diff
@@ -165,12 +165,13 @@ func (t *Net) Opts() []net.DialerOption {
...
 	if t.TLS != nil && t.TLS.Enabled {
 		cfg, err := tls.New(t.TLS.Opts()...)
-		if err == nil {
-			opts = append(opts,
-				net.WithTLS(cfg),
-			)
+		if err != nil {
+			return nil, err
 		}
+		opts = append(opts,
```
Comment (rinx) [inline internal/config/net.go:171]: Like in the internal/config/grpc.go (https://github.com/vdaas/vald/blob/refactor/internal/refactor-config-error-handling/internal/config/grpc.go#L244-L248), does it need to check `cfg` is not nil?

**Review by rinx — COMMENTED** [review]
Please rename this PR because it will be shown in the Changelog.

**Comment by rinx** [comment]
@vankichi about e2e bench workflows, there's no need to fix it immediately.

File: internal/config/net.go (hunk 3 above)
Comment (vankichi) [reply internal/config/net.go:171]: it seems to be not needed.
Maybe un-reaach code?
I'll remove it and update test if needed

Comment (vankichi) [reply internal/config/net.go:171]: sorry, I misread your comment.
the nil check is not must needed. sorry, my change is not consistent...
if we have to implement a politely code, nil check is better.

Comment (vankichi) [reply internal/config/net.go:171]: While searching, I found out every functional option `WithTLS(cfg)` (redis, cassandra, grpc, net) checks whether `cfg` is nil or not.
So, I decided to remove nil check from  https://github.com/vdaas/vald/blob/refactor/internal/refactor-config-error-handling/internal/config/grpc.go#L244-L248.

**Review by rinx — DISMISSED** [review]

**Review by kevindiu — DISMISSED** [review]
LGTM

**Review by hlts2 — DISMISSED** [review]

**Review by kpango — APPROVED** [review]
LGTM

**Pull request merged**